### PR TITLE
STOR-1310: Enhancement for Secrets Store CSI Driver Operator

### PR DESCRIPTION
**Pull request by dobsonj**
Epic: https://issues.redhat.com/browse/STOR-676
Story: https://issues.redhat.com/browse/STOR-1310

/cc @openshift/storage @gcharot @anjalitelang @lpettyjo @duanwei33
/assign @jsafrane @JoelSpeed

### CODE REVIEW

File: enhancements/storage/csi-secrets-store.md [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+3. The operator removes the CSI driver and associated manifests. Cluster admin verifies the CSI driver pods are no longer running.
+4. Cluster admin uninstalls the Secrets Store CSI driver operator via OLM. This should be done only after the CSI driver itself was already removed. Removing the operator does not automatically remove the driver.
+
+### API Extensions
+
+The only planned OpenShift API change is to add [secrets-store.csi.k8s.io](https://github.com/openshift/api/blob/bdd8865676216305be6065ade196f36752dfd015/operator/v1/types_csi_cluster_driver.go#L86) to the list of supported providers in the ClusterCSIDriver object.
+
+Once the operator is installed and ClusterCSIDriver object is created, the operator must apply two new CRD's required by the provider plugins:
+[secretproviderclasses](https://github.com/openshift/secrets-store-csi-driver/blob/main/deploy/secrets-store.csi.x-k8s.io_secretproviderclasses.yaml) and
+[secretproviderclasspodstatuses](https://github.com/openshift/secrets-store-csi-driver/blob/main/deploy/secrets-store.csi.x-k8s.io_secretproviderclasspodstatuses.yaml).
```
Comment (jsafrane) [inline enhancements/storage/csi-secrets-store.md:103]: We should mention (somewehere else?) that the operator installs  RBAC ClusterRoles so
* authenticated user can create/modify/delete SecretProviderClasses in their namespaces (~ something like reference to a Secret in the 3rd party secret store)
* authenticated user can read SecretProviderClassPodStatuses (~ extension of Pod.Status)

File: enhancements/storage/csi-secrets-store.md [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+### Implementation Details/Notes/Constraints [optional]
```
Comment (jsafrane) [inline enhancements/storage/csi-secrets-store.md:105]: Will OCP enable any of the advanced features like:
* Auto rotation of mounted contents
* Sync as Kubernetes Secret
* Set as ENV var

They may add more churn to the driver and should not be used in larger clusters.

File: enhancements/storage/csi-secrets-store.md [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+the driver can be used from `restricted` namespaces by default. The cluster admin
+can change this value if they wish to limit the use of this driver to `privileged`
+or `baseline` namespaces.
+
+### Publishing New Provider Plugins
+
+Refer to [upstream guidance for provider implementation](https://secrets-store-csi-driver.sigs.k8s.io/providers.html) when developing a new provider plugin. Each provider is expected to:
+- Use the functions and data structures in [service.pb.go](https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/main/provider/v1alpha1/service.pb.go) to implement the server code.
+- Run a DaemonSet that is deployed to the same nodes as the [CSI driver DaemonSet](https://github.com/openshift/secrets-store-csi-driver-operator/blob/main/assets/node.yaml).
+- Create a Unix Domain Socket under `/etc/kubernetes/secrets-store-csi-providers`, which the [CSI driver uses](https://github.com/openshift/secrets-store-csi-driver-operator/blob/d271705b2e56b20581dcc04514c9b307b67fbeec/assets/node.yaml#L39) to communicate with the provider.
```
Comment (jsafrane) [inline enhancements/storage/csi-secrets-store.md:158]: Create a Unix Domain Socket under `/etc/kubernetes/secrets-store-csi-providers` **on the host**.


BTW, do we want to spread usage of sockets in /etc? They belong to `/var`

**Review by JoelSpeed — COMMENTED** [review]
Can you explain for me how this differs from the KMS provider integration that K8s supports? https://kubernetes.io/docs/tasks/administer-cluster/kms-provider/

Is there any relation between the two? 

IIUC KMS integration still stores secrets in etcd and this stores the secrets outside of the cluster right?

Otherwise I have very little feedback on this enhancement right now. Thank you for starting this as tech preview

File: enhancements/storage/csi-secrets-store.md (hunk 2 above)
Comment (dobsonj) [reply enhancements/storage/csi-secrets-store.md:106]: I added a note to this section, both 'auto rotation' and 'sync as secret' will be enabled in OCP.
Set as ENV var is not really a separate feature, it just relies on 'sync as secret' to create a k8s secret that can be mounted as ENV var.

The feedback I got on auto rotation is that it's useful enough that it should be enabled by default. Before this operator goes GA, we should have a way to toggle it on/off and set the polling interval though.

File: enhancements/storage/csi-secrets-store.md (hunk 3 above)
Comment (dobsonj) [reply enhancements/storage/csi-secrets-store.md]: Updated. The CSI driver actually looks at both paths, but `/var/run/secrets-store-csi-providers` is the one that should be documented. The [Azure provider](https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/dad67ff93e76554a489f72f2d4b0bef9afe2d417/deployment/provider-azure-installer.yaml#L72) we used in testing for example uses `/var`.

**Comment by JoelSpeed** [comment]
I had a look at the epic, looks like this has basically already been implemented and so the enhancement now I'm guessing is documenting what is done rather than what will be done, is that the correct interpretation?

**Pull request merged**